src/network.cpp

```diff
@@ -0,0 +1,1106 @@
+                 callback = std::move(handle_response)](
+                        bool success,
+                        bool timeout,
+                        int16_t status_code,
+                        std::optional<std::string> response,
+                        service_node_changes) {
+                    if (!success || timeout ||
+                        !ResponseParser::response_long_enough(builder.enc_type, response->size()))
+                        return handle_errors(status_code, response, info, callback);
+
+                    if constexpr (std::is_same_v<Destination, SnodeDestination>)
+                        process_snode_response(builder, *response, info, callback);
+                    else if constexpr (std::is_same_v<Destination, ServerDestination>)
+                        process_server_response(builder, *response, info, callback);
+                    else
+                        callback(false, false, -1, "Invalid destination.", service_node_changes{});
+                });
+    } catch (const std::exception& e) {
+        handle_response(false, false, -1, e.what(), service_node_changes{});
+    }
+}
+
+void Network::handle_errors(
+        const int16_t status_code,
+        const std::optional<std::string> response,
+        const request_info info,
+        network_response_callback_t handle_response) {
+    switch (status_code) {
+        // A 404 or a 400 is likely due to a bad/missing SOGS or file so
+        // shouldn't mark a path or snode as invalid
+        case 400:
+        case 404:
+            return handle_response(false, false, status_code, response, service_node_changes{});
+
+        // The user's clock is out of sync with the service node network (a
+        // snode will return 406, but V4 onion requests returns a 425)
+        case 406:
+        case 425:
+            return handle_response(false, false, status_code, response, service_node_changes{});
+
+        // The snode is reporting that it isn't associated with the given public key anymore. If
+        // this is the first 421 then we want to try another node in the swarm (just in case it was
+        // reported incorrectly). If this is the second occurrence of the 421 then the client needs
+        // to update the swarm (if the response contains updated swarm data), or increment the path
+        // failure count.
+        case 421:
+            try {
+                // If there is no response data or no swarm informaiton was provided then we should
+                // just replace the swarm
+                if (!info.swarm)
+                    throw std::invalid_argument{"Unable to handle redirect."};
+
+                // If this was the first 421 then we want to retry using another node in the swarm
+                // to get confirmation that we should switch to a different swarm
+                if (!info.is_retry) {
+                    std::random_device rd;
+                    std::mt19937 g(rd());
```

Comment:
Going out to the system random device isn't all that desirable (it can, for instance, block if called too much).  Too alternative approaches:
- copy CSRNG out of lokinet (from crypto/crypto.hpp).  Then you can just use it as the RNG instead of `mt19937`, i.e.:
```C++
CSRNG rng;
std::shuffle(swarm_copy.begin(), swarm_copy.end(), rng);
```
(and you get cryptographically secure random via libsodium, though that isn't needed here).

Option 2:
Make a global, thread_local `std::mt19937_64` that gets seeded once from `std::random_device` in its construction, then just use that global RNG here (and anywhere else that needs a non-secure random) instead of `g`.  oxen-core has an example of doing this in src/common/random.{h,cpp}.

Reply:
I ended up pulling in `CSRNG` and going with that approach as I needed to add a few more `shuffle` calls after pulling in the snode pool and path building logic so figured it may as well all be secure and consistent